---
title: "Differential Expression Analysis and Preliminary ORA on COVID-19 Signature Genes"
author: "Feifei Li"
date: "`r Sys.Date()`"
output:
    html_document:
        df_print: paged
        toc: true
        toc_depth: 2
        toc_float:
            collapsed: false
            smooth_scroll: true
        number_sections: true
bibliography: citations.bib
csl: biomed-central.csl
---


\newcommand{\sgn}{\operatorname{sgn}}
\newcommand{\fdr}{\mathrm{FDR}}
\newcommand{\fc}{\mathrm{FC}}


```{r data, include=FALSE}
source("./scripts/GSE152641_data_processing.R", local = knitr::knit_global())
source("./utils/plotters.R", local = knitr::knit_global())
source("./utils/gprofiler_wrapper.R", local = knitr::knit_global())
```


# Introduction to the dataset used

The original dataset contains RNA-seq count data generated by profiling  peripheral blood from 62 COVID-19 patients (**COVID-19** group) and 24 healthy controls (**HC** group) via bulk RNA-seq. It is publicly available on Gene Expression Omnibus database (GEO) under accession number  [GSE152641](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152641)[@GSE152641] as a count matrix with 20460 unique Entrez gene ID's along the rows and 86 samples (biological replicates) along the columns.
The original study not only performed expression analysis on genes of these 86 samples, but also compiled expression data of prior studies on six viruses: influenza, RSV, HRV, Ebola, Dengue, and SARS in an attempt to isolate COVID-19 biomarkers from other viral infections. Here we will only perform an expression analysis for the RNA-seq count matrix provided by this study.
The count matrix file was downloaded using [GEOquery](https://bioconductor.org/packages/release/bioc/html/GEOquery.html)[@GEOquery] version 2.55.1.
TMM normalisation was applied to the data using [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)[@edgeR_1][@edgeR_2] version 3.30.0 before exploratory data analysis. During the normalisation process, 6035 genes had expression values that met the criterium of being outliers as per the edgeR protocol[@edgeR_protocol] and hence were removed.
The mapping from Entrez ID against the official gene symbol (HUGO/HGNC) of the genes was done using [annotate](https://bioconductor.org/packages/release/bioc/html/annotate.html)[@gentleman2017annotate] and database [org.Hs.eg.db](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html)[@carlson2017org]. The mapping resulted in one duplicate record of gene **ZFHX3**, to which the gene of reference ID 388289 and gene 463 were both mapped:

| Entrez ID | Gene Symbol |                     Status                     |
|:---------:|:-----------:|:----------------------------------------------:|
|   388289  |   C16orf47  | Replaced with Gene ID: 463, HUGO symbol: ZFHX3 |

Since the expression values for these two ID's are identical, and ID 388289 is obsolete, I decided to remove the record for ID 388289 from the normalised expression matrix; retaining it could introduce noises in further analysis.

For the exploratory analysis, [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)[@edgeR_1][@edgeR_2] was also used to produce a multidimensional scaling (MDS) plot, a Biological Coefficient of Variation (BCV) plot, and a mean-variance plot to present evidence of differences between our biological replicates of the two groups (HC and COVID-19).

```{r, warning=FALSE, message=FALSE, echo=FALSE}
edgeR::plotMDS.DGEList(
    gene_counts_DGE,
    col      = c("#b7371d", "#51af84")[factor(samples_by_group$group)],
    cex.axis = 0.75,
    cex.lab  = 0.75,
    main     = paste("Variation between Samples",
                     "in Expression Level")
    )
legend(x        = -2.05,
       y        = 1.95,
       lwd      = 2,
       text.col = c("#51af84","#b7371d"),
       legend   = c("HC", "COVID-19"),
       lty      = 0,
       cex      = 0.75)
```


The MDS plot presents variation amongst samples based on the normalised gene expression. The distances between each pair of samples on the MDS plot suggest their leading logFC. The healthy control (HC) and the patients (COVID-19) samples form separate clusters along the plot dimension 1. There are however still several COVID-19 samples mixed in the HC cluster.



```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Regression model with group as target and samples as predictors
model_design <- model.matrix(~group, data  = samples_by_group)
dispersion   <- edgeR::estimateDisp(y      = gene_counts_DGE,
                                    design = model_design)
edgeR::plotBCV(
    y           = dispersion,
    col.tagwise = "black",
    col.common  = "red",
    main        = paste("Common and Tag-wise Dispersion vs",
                        "Expression Levels")
)
```



The BCV plot presents the estimated relative variability of expression between biological replicates; it illustrates the association between the biological CV and the average true gene abundance. The common dispersion line suggests all gene expression values vary by close to a BCV value of 0.5 amongst replicates.[@bcv]
The tagwise dispersion shows BCV values calculated individually for each gene. We observe that genes with higher true abundance (under the assumption that RNA-seq counts follow a negative binomial distribution) have lower BCV's than genes with lower abundance. This imples that the higher the true expression level of a gene has, the lower its variation in our samples.


```{r, warning=FALSE, message=FALSE, echo=FALSE}
edgeR::plotMeanVar(
    object                       = dispersion,
    show.raw.vars                = TRUE,
    show.tagwise.vars            = TRUE,
    NBline                       = TRUE,
    show.ave.raw.vars            = TRUE,
    show.binned.common.disp.vars = TRUE,
    main                         = paste(
        "Mean-variance Relationship for 14425 Genes of",
        "COVID-19 Group and Healthy Control Group"),
    cex.main                     = 0.8
)
```


The mean-variance plot presents the modelling of the mean-variance relationship for the normalised expression values, which are split into bins by overall expression level in the model. The grey points represent the pooled genewise variances. The blue points on the mean-variance plot represent the estimated genewise variances. The red crosses represent the pooled genewise variances, while the dark red crosses represent the the average of the pooled variances for each bin of genes plotted against the average expression level of the genes in the bin. The blue line shows the mean-variance relationship for a negative binomial model with common dispersion.[@edgeR_1][@edgeR_2] We obeserve that all types of variances fit well along the negative binomial line.


Note that this result can only be reproduced using *Bioconductor 3.11*; using other versions of Bioconductor will lead to a different mapping of gene names.

# Differential Gene Expression

## Visualise with Heatmap: First Attempt

We will be using [ComplexHeatmap](https://bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html)[@complexheatmap] for heatmap visualisation of gene expression and [circlize](https://github.com/jokergoo/circlize)[@circlize] to generate a colour gredient indicating the expression levels of genes.

Let's try to plot out all the genes in a heatmap to see if there is any noticible pattern:
```{r, warning=FALSE, message=FALSE, echo=FALSE}
plotHeatMap(m       = gene_counts_norm,
            title   = paste("Expression Levels of All",
                            as.character(nrow(gene_counts_norm)),
                            "Genes"),
            samples = samples_by_group)
```



We observed no signs of differential expression.

## Differential Expression Analysis by Quasi-Likelihood Methods (QLM)

### Model Design

We hypothesized that COVID-19 status of samples was the only factor contributing to differential gene expression. Therefore, we model on *group* (status COVID-19 or HC) such that fitting this design matrix will tell us how the COVID-19 status of a sample explains his/her expression levels of genes.
```{r, warning=FALSE, message=FALSE}
(model_design <- model.matrix(~group, data = samples_by_group))
```



Note that *samples_by_group* is a dataframe labeling samples with their groups:
```{r, warning=FALSE, message=FALSE}
samples_by_group
```


### Multiple Hypothesis Testing for Differentially Expressed Genes by QLM
We first use edgeR[@edgeR_1][@edgeR_2] to estimate the dispersion. Note that the object *d* is a DGEList previously created when normalising the count matrix (see the data processing script for details).
```{r, warning=FALSE}
d       <- edgeR::estimateDisp(d, model_design) # dispersion of normalised count
fit_qlm <- edgeR::glmQLFit(d, model_design)
```
In this step, QL dispersion was estimated with edgeR[@edgeR_1][@edgeR_2] by fitting a QL negative binomial (NB) generalised log-linear model to the DGEList data containing the normalised counts and the estimated NB dispersions (visualised with the BCV and mean-variance plots), along with the design matrix we just created.


We then use the fitted result to conduct genewise QL F-test for the coefficients of our defined sample groups.
```{r}
qlf_SARS2vsHC <- edgeR::glmQLFTest(fit_qlm, coef = colnames(model_design)[2])
```
This step calculates the p-values for each of the genes in our expression set.

### QL F-test P-values Comparison: FDR-correcred vs Uncorrected
Extract the top differentially expressed tags ranked by p-values from the result of the QL F-test and display the number of genes that passed the correction:
```{r, echo=FALSE}
qlf_output_hits <- edgeR::topTags(object         = qlf_SARS2vsHC,
                                  sort.by        = "PValue",
                                  n              = nrow(gene_counts_norm))
knitr::kable(data.frame(
    "Number of genes with p-value < 0.05"        = length(
        which(
            qlf_output_hits$table$PValue < 0.05
            )
        ),
    "Number of genes that pass after correction" = length(
        which(
            qlf_output_hits$table$FDR < 0.05
        )
    ),
    check.names                                  = F
))
```


The threshold for showing statistically significant evidence for differential expression was set to $p < 0.05$ for each gene. We chose this threshold because we want to capture the genes that only have less 5% chance to show such differences in expression between groups if they were non-differentially expressed genes.


Note that edgeR by default uses the false discovery rate (FDR/Benjamini-Hochberg) method to correct p-values for false positive results. This method was applied because we need to control for the the liklihood of false positive results that would increase by chance with the increasing number of tests performed.
We also set the threshold for the corrected p-value (FDR) to be $<0.05$, because we want to capture genes that have false positive results for fewer than 5% of the significant tests.

### Volcano Plot for QLF Results

```{r, warning=FALSE, message=FALSE, echo=FALSE}
qlf_volcano <- data.frame(
    gene       = entrez_to_gname[rownames(qlf_output_hits$table)],
    log2FC     = qlf_output_hits$table$logFC,
    neg_log10P = -log10(qlf_output_hits$table$FDR)
)
plotVolcano(
    df = qlf_volcano,
    title = "QL F-test Results for 14425 Genes (HC vs COVID-19)",
    gene_of_interest = c("ACO1", "ATL3", "ZC3H13")
)
```


A Volcano plot was chosen over an MA plot to show the amount of differentially expressed genes because a volcano plot can answer such questions as how many genes that passed the correction are up-regulated or are down-regulated, while an MA plot fails to visualise the association between the results of multiple hypothesis testing and differential expression.
Note that the p-values we used for the volcano plot are the FDR-corrected p-values.


Here we highlighted two genes of interest: **ACO1** and **ATL3**. These two genes were identified by the original study to be the most important **COVID-19 signature genes**. Thair .et al discovered that ACO1 and ATL3 are completely oppositely regulated between COVID-19 patients and non-COVID-19 patients; ACO1 is up-regulated in COVID-19 patients and down-regulated in patients with non-COVID-19 viral infections, while ATL3 is down-regulated in COVID-19 patients and up-regulated in non-COVID-19 patients. The authors also found that **ZC3H13** is the most down-regulated genes amongst the COVID-19 signature genes.[@GSE152641] The position on our volcano plot reflect that the result of our differential expression analysis by QLM coincides with this observation made by Thair .et al; ACO1 is down-regulated while ATL3 is up-regulated in the healthy control.

### Heatmap for QLF top-hit Genes

```{r, warning=FALSE, message=FALSE, echo=FALSE}
tophits_qlf <- entrez_to_gname[
    rownames(qlf_output_hits$table[
        which(qlf_output_hits$table$FDR < 0.05)
    , ])
]
plotHeatMap(tophits = tophits_qlf,
           m       = gene_counts_norm,
           samples = samples_by_group,
           title   = "Expression Levels of QLF Top-hit Genes")
```


The heatmap plotted from the top-hit genes ($FDR < 0.05$) from the differential expression analysis exhibits clusters of up-regulated and down-regulated expressions between the two group. This was not observed in the heatmap plotted using all the genes because here we have removed genes that show no significant evidence for differential expression for COVID-19 status ($FDR \geq 0.05$), which introduced noises in the first heatmap.


## Differential Expression Analysis by Linear Models of MircroArray (Limma)

### Multiple Hypothesis Testing

We are fitting the same model matrix as the one we used for QLM fitting:
```{r, warning=FALSE, message=FALSE}
fit_limma <- limma::lmFit(gene_counts_norm, design = model_design)
```
Then we apply empircal Bayes to compute differential expression:
```{r, warning=FALSE, message=FALSE}
fit_limma <- limma::eBayes(
    fit = fit_limma,
    trend = T # specific to RNA-seq, as we are not working with microarray data
)
```

Get the top hits ranked by P-values:
```{r, warning=FALSE, message=FALSE}
limma_output_hits <- limma::topTable(
    fit = fit_limma,
    coef = which(colnames(fit_limma$coefficients) == "groupHC"),
    adjust.method = "BH", # use Benjamni-Hochberg to correct for p-values
    number = nrow(gene_counts_norm)
    )
limma_output_hits <- limma_output_hits[order(limma_output_hits$P.Value), ]
```

### P-value adjustment methods

For the results of empircal Bayes, we are able to select whatever p-value correction mathods R offers:
```{r, warning=FALSE, message=FALSE}
p.adjust.methods
```
Note that *BH* and *fdr* are synonymous; they both refer to the Benjamini-Hochberg procedure for FDR control. According to R documentation for these adjustment methods, *holm*, *hochberg*, *hommel*, and *bonferroni* are designed to control the family-wise error rate (the probability of making one or more false postive discoveries; Type 1 error) and they control the probability of making such error in the multi-hypotheses testing. *BH/fdr* and *BY* on the other hand allow false positive results, but control the proportion out of the significant tests. This means the latter two methods are less conservative and stringent than the first four. Although *hommel* is a more powerful method, the R documentation states that the difference between the results of these two methods are small while the p-values of *BH* are faster to calculate. Therefore the FDR method was also chosen here.


### Limma P-values Comparison: BH-adjusted vs Unadjusted

Check the number of genes that have P-values below the 0.05 cutoff and
genes that have Benjamni-Hochberg corrected P-values below the cutoff:
```{r, echo=FALSE, message=FALSE}
knitr::kable(data.frame(
    "Number of genes with p-value < 0.05"        = length(
        which(
            limma_output_hits$P.Value < 0.05
            )
        ),
    "Number of genes that pass after correction" = length(
        which(
            limma_output_hits$adj.P.Val < 0.05
        )
    ),
    check.names                                  = F
))
```
The result is quite similar to that of the QLM F-test, with slightly fewer genes that passed the threshold.

### Volcano Plot for Limma top-hit Genes

```{r, warning=FALSE, message=FALSE, echo=FALSE}
limma_volcano <- data.frame(
    gene       = rownames(limma_output_hits),
    log2FC     = limma_output_hits$logFC,
    neg_log10P = -log10(limma_output_hits$adj.P.Val)
)
plotVolcano(
    df               = limma_volcano,
    title            = "Limma-eBayes Results for 14425 Genes (HC)",
    gene_of_interest = c("ACO1", "ATL3", "ZC3H13")
)
```



The volcano plot behaves oddly for the limma result because of an outlier showing expression of -5000 fold change, and an outlier showing expression of higher than 1000 fold change; this could be attributed to the different methods edgeR and limma used to calculate log2 fold change. Genes cluster around the zero fold change level, and it becomes difficult to tell whether the signature genes are up- or down-regulated here.


### Heatmap for Limma top-hit Genes

```{r, warning=FALSE, message=FALSE, echo=FALSE}
tophits_limma <- rownames(limma_output_hits[which(
    limma_output_hits$adj.P.Val < 0.05
    ), ])
plotHeatMap(tophits = tophits_limma,
            m       = gene_counts_norm,
            title   = "Expression Levels of Limma Top-hit Genes",
            samples = samples_by_group)
```



The heatmap for the limma top-hit genes has the same pattern of clusters as the one for the QLF top-hit genes. This again is because we only retain genes that have are significant for the differential expression analysis by limma-eBayes, and ploted the two groups side-by-side for comparison.


## Comparing QL F-test and Limma Results

```{r, include=FALSE}
qlf_pvalues     <- data.frame(
    gene_name      = entrez_to_gname[rownames(qlf_output_hits$table)],
    qlf_pvalue     = qlf_output_hits$table$PValue,
    qlf_pvalue_adj = qlf_output_hits$table$FDR
    )
limma_pvalues     <- data.frame(
    gene_name        = rownames(limma_output_hits),
    limma_pvalue     = limma_output_hits$P.Value,
    limma_pvalue_adj = limma_output_hits$adj.P.Val
    )
qlf_limma_pvalues <- merge(
    x    = qlf_pvalues,
    y    = limma_pvalues,
    by.x = "gene_name",
    by.y = "gene_name",
)
qlf_limma_pvalues$colour <- "grey"
```

```{r, warning=FALSE, message=FALSE, echo=FALSE}
par(mfrow = c(1, 2))
# Scatter plots for un-adjusted p-values ======================================
# Colour for QLF p-values below cutoff
qlf_limma_pvalues$colour[
    which(qlf_limma_pvalues$qlf_pvalue < 0.05)] <- "#92d964"
# colours for limma p-values below cutoff
qlf_limma_pvalues$colour[
    which(qlf_limma_pvalues$limma_pvalue < 0.05)] <- "#29a2ca"
# Colour for intersection of QLF and limma p values below cutoff
qlf_limma_pvalues$colour[which(
    qlf_limma_pvalues$qlf_pvalue   < 0.05 &
    qlf_limma_pvalues$limma_pvalue < 0.05)] <- "#fae213"
plot(
    x = qlf_limma_pvalues$limma_pvalue,
    y = qlf_limma_pvalues$qlf_pvalue,
    col = qlf_limma_pvalues$colour,
    xlab = "Limma Unadjusted P-values",
    ylab = "QLF Unadjusted P-values",
    cex.lab = 0.75,
    main = "QLF vs Limma for Unadjusted P-values",
    cex.main = 0.75
)
legend(
    x = 0, y = 1,
    legend = c("QLF", "Limma", "Both", "not significant"),
    fill = c("#92d964", "#29a2ca", "#fae213", "grey"),
    cex = 0.5
)
qlf_limma_pvalues$colour <- "grey"
# Scatter plots for adjusted p-values =========================================
# Colour for FDR-corrected QLF p-values below cutoff
qlf_limma_pvalues$colour[which(
    qlf_limma_pvalues$qlf_pvalue_adj < 0.05)] <- "#168c6e"
# Colour for BH-adjusted limma p-values below cutoff
qlf_limma_pvalues$colour[which(
    qlf_limma_pvalues$limma_pvalue_adj < 0.05)] <- "#234eaa"
# Colour for intersection of corrected QLF and limma p values below cutoff
qlf_limma_pvalues$colour[which(
    qlf_limma_pvalues$limma_pvalue_adj < 0.05 &
    qlf_limma_pvalues$qlf_pvalue_adj < 0.05)] <- "#cb641d"
plot(
    x        = qlf_limma_pvalues$limma_pvalue_adj,
    y        = qlf_limma_pvalues$qlf_pvalue_adj,
    col      = qlf_limma_pvalues$colour,
    xlab     = "Limma BH-corrected P-values",
    ylab     = "QLF FDR-corrected P-values",
    cex.lab  = 0.75,
    main     = "QLF vs Limma for Adjusted P-values",
    cex.main = 0.75
)
legend(
    x        = 0,
    y        = 1,
    legend   = c("QLF", "Limma", "Both", "not significant"),
    fill     = c("#168c6e", "#234eaa", "#cb641d", "grey"),
    cex      = 0.5
)
```


The p-values calculated by QLF and Limma methods exhibit a strong positive linear correlation, suggesting the similarity in the results of these two differential expression analyses. Moreover, both these two methods can capture a common subset of top hits; this explains the nearly identical clustering pattern in the two heatmaps.


# Thresholded Over-Representation Analysis (ORA)

The thresholded ORA was performed using the top-hit differentially expressed genes from the **QLF** result. This is because we couldn't confirm whether the signature genes ACO1 and ATL3 are up- or down-regulated in HC from the volcano plot for the limma result and compare against the observation made by Thair .et al. The unusally high and low fold changes of some genes in the limma result also makes me sceptical of its credibility.

## ORA Method

To perform thresholded ORA, we used [g:Profiler](https://cran.r-project.org/web/packages/gprofiler2/vignettes/gprofiler2.html)[@gprofiler]. The reasons why we choose g:Profiler here is that, first, it provides R interface, while many of other ORA tools are only availabel as web servers; second, compared to the other popular R package *clusterProfiler* for enrichment analysis, g:Profiler allows users to query multiple annotation sources in one query and combine the results, while we could only query single annotation source one at a time with clusterProfiler. Although clusterProfiler has a set of more versatile plotting functions than g:Profiler R interface does, the out-of-box interactive Manhattan plot combining the results of different annotation sources is a better option for preliminary ORA.


## Annotation Data

Since this is a pathway analysis, we are only interested in data sources for biological pathways. Hence, **Reactome**, **Wikipathways**, and **KEGG** were included. Although sometimes KEGG returns uninformative pathways, we still included it here to get a more comprehensive result. We will also include **GO:Biological process (GO:BP)** as our data source. Note that a biological process is different from a pathway; however, pathways are considered to collectively participate in biological processes. Therefore, we also included GO:BP as one of our data sources.

The databases used by the current release of g:Profiler are *Ensembl 102*, which is also the current version of GO database, and *Ensembl Genomes 49*.
The currect release of Reactome used is *version 75*. Wikipathways is a comunity-maintained database and is being constantly updated. The current release of KEGG used is *97.0*.


## Threshold, P-value Correction, and IEA

The p-value threshold for the significance of the gene-set enrichment of the Fisher's exact test calculated by g:Profiler was set to $0.05$. The numbers of genesets that were returned with this threshold by each query would be presented as a summary below. We chose Benjamini–Hochberg FDR for p-value adjustment method, and exclude IEA results (Inferred from Electronic Annotation) since we want the human-curated results, which have better quality.

## Thresholded Lists

We need to create 2 thresholded lists: one for the top-hit genes that were up-regulated, and one for those downregulated.


<!-- ### Thresholded List for Upregulated Top-hit Genes -->

```{r threshold_up, warning=FALSE, message=FALSE}
upregulated_genes <- entrez_to_gname[rownames(qlf_output_hits$table[which(
    qlf_output_hits$table$FDR < 0.05 & qlf_output_hits$table$logFC > 0
),])]
```

<!-- ### Thresholded List for Downregulated Top-hit Genes -->

```{r threshold_down, warning=FALSE, message=FALSE}
downregulated_genes <- entrez_to_gname[rownames(qlf_output_hits$table[which(
    qlf_output_hits$table$FDR < 0.05 & qlf_output_hits$table$logFC < 0
),])]
```

An overview of our top-hit genes by QLF:
```{r tophit_overview, warning=FALSE, message=FALSE, echo=FALSE}
knitr::kable(data.frame(
    "Upregulated top-hit genes"     = length(which(
        qlf_output_hits$table$FDR   < 0.05 &
        qlf_output_hits$table$logFC > 0
    )),
    "Downregulated top-hit genes"   = length(which(
        qlf_output_hits$table$FDR   < 0.05 &
        qlf_output_hits$table$logFC < 0
    )),
    check.names                     = F
))
```

## All Thresholded Differentially Expressed Genes

<!-- ### Building g:Profiler query: -->

```{r gost_all, message=FALSE, warning=FALSE, echo=FALSE}
gostres_tophits <- gostQuery(genelist = tophits_qlf)
```

### Query Summary

```{r, message=FALSE, warning=FALSE, echo=FALSE}
knitr::kable(summariseGost(gostres = gostres_tophits))
```
Note that the threshold for the query is $p < 0.05$.

### Graphical Summary

```{r, warning=FALSE, message=FALSE, echo=FALSE}
gprofiler2::gostplot(gostres = gostres_tophits)
```

### Query Result Detail

There are too many pathways to show, and since GO terms are propagated,
there might be terms that are too board and not so informative.
Hence we limit the term size to fewer than 1000 genes and
only look at those that passed p-value threshold and FDR correction:


#### GO Biological Process
```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_tophits, upper = 1000, source = "GO:BP")
```

#### Reactome

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_tophits, upper = 1000, source = "REAC")
```


#### WikiPathways

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_tophits, upper = 1000, source = "WP")
```


#### KEGG

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_tophits, upper = 1000, source = "KEGG")
```

## Upregulated Genes

<!-- ### Building g:Profiler query: -->
```{r gost_up, message=FALSE, warning=FALSE, echo=FALSE}
gostres_up <- gostQuery(genelist = upregulated_genes)
```

### Query Summary

```{r, message=FALSE, warning=FALSE, echo=FALSE}
knitr::kable(summariseGost(gostres = gostres_up))
```

### Graphical Summary

```{r, warning=FALSE, message=FALSE, echo=FALSE}
gprofiler2::gostplot(gostres = gostres_up)
```

### Query Result Detail

#### GO Biological Process
```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_up, upper = 200, source = "GO:BP")
```

#### Reactome

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_up, upper = 200, source = "REAC")
```


#### WikiPathways

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_up, upper = 200, source = "WP")
```

#### KEGG

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_up, upper = 1000, source = "KEGG")
```

## Downregulated Genes

The workflow is the same as the enrichment analysis for the upregulated genes.

<!-- ### Building g:Profiler query: -->

```{r gost_down, message=FALSE, warning=FALSE, echo=FALSE}
gostres_down <- gostQuery(genelist = downregulated_genes)
```

### Query Summary

```{r, message=FALSE, warning=FALSE, echo=FALSE}
knitr::kable(summariseGost(gostres = gostres_down))
```

### Graphical Summary

```{r, warning=FALSE, message=FALSE, echo=FALSE}
gprofiler2::gostplot(gostres = gostres_down)
```

### Query Result Detail

#### GO Biological Process
```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_down, upper = 200, source = "GO:BP")
```

#### Reactome

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_down, upper = 200, source = "REAC")
```

#### WikiPathways

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_down, upper = 200, source = "WP")
```

#### KEGG

```{r, warning=FALSE, message=FALSE, echo=FALSE}
termTable(gostres = gostres_down, upper = 1000, source = "KEGG")
```

## ORA Result Comparison

*Run the analysis using the up-regulated set of genes, and the down-regulated set of genes separately. How do these results compare to using the whole list (i.e all differentially expressed genes together vs. the up-regulated and down regulated differentially expressed genes separately)?*

The number of gene sets returned appears to be proportional to the length of the query gene list. All these three results using three different gene lists reveal different information.

In terms of GO biological processes, the result for all differentially expressed genes involve regulation of *neutrophil*, *leukocyte*, and *myeloid cells*. The result for up-regulated genes does not seem to have many informative GO terms except for viral gene expression and viral transcription. The result for the down-regulated genes has terms involve *type I interferon* and defense response to virus.

In terms of REACTOME, pathways involving *neutrophil* regulation and *myeloid leukocyte* also appear in the result for all differentially expressed genes. The result for the up-regulated genes does not seem to have many informative terms either. The result for the down-regulated genes also shows signaling pathways for *interferon*.

In terms of wikipathways, the results for all differentially expressed genes shows *T-Cell antigen Receptor (TCR) Signaling Pathway* and *Type II interferon signaling (IFN)*; pathways involving type I interferon appear too. The result for the up-regulated genes also have pathways involving *TCR*. The result for the down-regulated genes is similar to the previous two sources, showing pathways involving *Type I interferon* and *Type II interferon*.

In terms of KEGG, the result for all differentially expressed genes has "Coronavirus disease - COVID-19" has the top hit pathway, expected but not informative; many cancer-relative pathways also appear. The result for the up-regulated genes shows a pathway involving TCR. The KEGG pathways for the down-regulated genes are mostly cancer-related.

# Overall Interpretation

*Do the over-representation results support conclusions or mechanism discussed in the original paper?*


The ORA results highly support the conclusions drawn in the original paper.
The authors also found pathways such as neutrophil activation, innate immune response,
immune response to viral infection, type-I interferon signaling,
and cytokine production for 771 upregulated genes they discovered in COVID-19 patients;
pathways they found for 1231 down-regulated genes include
lymphocyte differentiation and T-cell activation and regulation,
which also appear in our ORA result. The authors concluded that
T-cell are suppressed while neutrophils are activated as
an indicator of host response to COVID-19 represented in the transcriptomic changes.[@GSE152641]


*Can you find evidence, i.e. publications, to support some of the results that you see. How does this evidence support your results.*


There have been prior researches on host immune responses to COVID-19 showing findings that support this enrichment analysis result.
An immune analysis on COVID-19 patients by Hadjadj et al. has shown findings that coincide with our ORA result.
Their multiplex gene expression analysis showed
an up-regulation of genes involved in type I interferon signaling in COVID-19 patients[@Hadjadj718],
while pathways involving type I interferon are present in our ORA results for down-regulated genes in healthy control.
Anurag et al. observed increased total leukocyte count and differential neutrophil count in patients with severe COVID-19 infection[@anurag2020],
and GO terms *leukocyte activation involved in immune response* and *neutrophil activation involved in immune response* are present in the ORA result for all differentially expressed genes.
Kalfaoglu B. et al. discovered a unique dynamics of T-cells in severe COVID-19 patients: T-cells become hyperactivated, proliferate and die rapidly before differentiating into Treg, while they have shown that the majority of Treg-type genes are regulated by TCR signaling[@hyper_Tcell]; this supports the involvement of T-cell signaling pathways in our ORA result.


# Note

## Normalisation Script

For details regarding the processing of the data, see *GSE152641_data_processing.R* under *scripts* directory of the repository.

## Helper Functions

For details regarding implementation of helper functions used in this markdown file, see directory *utils*.

## Ranked Genelist

Although we are performing thresholded ORA, and GESA is not in the scope of this assignment, a ranked gene-list was asked at the beginning of the assignment description and is provided below. A data file will also be created in */data* directory once the Rmd file is compiled. Note that it will not be used for the ORA as it is not the appropriate tool for a ranked/un-thresholded gene-list.

```{r rank_list, warning=FALSE, message=FALSE}
gene_rank <- data.frame(
    genename = entrez_to_gname[rownames(qlf_output_hits$table)],
    F_stat   = -log(qlf_output_hits$table$FDR,
                    base = 10) * sign(qlf_output_hits$table$logFC)
)
gene_rank
```

Note that we define the rank to be $-\log_{10}{\fdr} \ \cdot\ \sgn(\log{\fc})$. With the $-\log_{10}{\fdr}$, the samller the (corrected) p-value for a gene is, the higher the rank of that gene, regardless of whether it is up- or downregulated. The other term $\sgn(\log{\fc})$ take into account the factor of regulation of gene expression:
if it is up-regulated, then it will be ranked from the top;
if it is down-regulated, then it will be ranked from the bottom.
Thereby we have genes that show the most statistically significant evidence of differential expression at the top for those upregulated, and the bottom of the list for those downregulated, while genes in the middle are least significant. Again, gene enrichment analysisfor the non-thresholded ranked list is not in the scope of this assignment and hence will not be run.

```{r, include=FALSE}
write.table(x         = gene_rank,
            file      = file.path(getwd(),
                             "data",
                             "covid19_ranked_genelist.txt"),
            sep       = "\t",
            row.names = F,
            col.names = F,
            quote     = F)
```


# References
