---
title: "Dataset"
author: "Feifei Li"
date: "`r Sys.Date()`"
output:
    html_document:
        df_print: paged
---

```{r, include=FALSE}
source("../scripts/GSE152641_data_processing.R",
       chdir = TRUE,
       local = knitr::knit_global())
```

The original dataset contains RNA-seq count data generated by profiling  peripheral blood from 62 COVID-19 patients (**COVID-19** group) and 24 healthy controls (**HC** group) via bulk RNA-seq. It is publicly available on Gene Expression Omnibus database (GEO) under accession number  [GSE152641](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE152641)[@GSE152641] as a count matrix with 20460 unique Entrez gene ID's along the rows and 86 samples (biological replicates) along the columns.
The original study not only performed expression analysis on genes of these 86 samples, but also compiled expression data of prior studies on six viruses: influenza, RSV, HRV, Ebola, Dengue, and SARS in an attempt to isolate COVID-19 biomarkers from other viral infections. Here we will only perform an expression analysis for the RNA-seq count matrix provided by this study.
The count matrix file was downloaded using [GEOquery](https://bioconductor.org/packages/release/bioc/html/GEOquery.html)[@GEOquery] version 2.55.1.
TMM normalisation was applied to the data using [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)[@edgeR_1][@edgeR_2] version 3.30.0 before exploratory data analysis. During the normalisation process, 6035 genes had expression values that met the criterium of being outliers as per the edgeR protocol[@edgeR_protocol] and hence were removed.
The mapping from Entrez ID against the official gene symbol (HUGO/HGNC) of the genes was done using [annotate](https://bioconductor.org/packages/release/bioc/html/annotate.html)[@gentleman2017annotate] and database [org.Hs.eg.db](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html)[@carlson2017org]. The mapping resulted in one duplicate record of gene **ZFHX3**, to which the gene of reference ID 388289 and gene 463 were both mapped:

| Entrez ID | Gene Symbol |                     Status                     |
|:---------:|:-----------:|:----------------------------------------------:|
|   388289  |   C16orf47  | Replaced with Gene ID: 463, HUGO symbol: ZFHX3 |

Since the expression values for these two ID's are identical, and ID 388289 is obsolete, I decided to remove the record for ID 388289 from the normalised expression matrix; retaining it could introduce noises in further analysis.

For the exploratory analysis, [edgeR](https://bioconductor.org/packages/release/bioc/html/edgeR.html)[@edgeR_1][@edgeR_2] was also used to produce a multidimensional scaling (MDS) plot, a Biological Coefficient of Variation (BCV) plot, and a mean-variance plot to present evidence of differences between our biological replicates of the two groups (HC and COVID-19).

```{r, warning=FALSE, message=FALSE, echo=FALSE}
edgeR::plotMDS.DGEList(
    gene_counts_DGE,
    col      = c("#b7371d", "#51af84")[factor(samples_by_group$group)],
    cex.axis = 0.75,
    cex.lab  = 0.75,
    main     = paste("Variation between Samples",
                     "in Expression Level")
    )
legend(x        = -2.05,
       y        = 1.95,
       lwd      = 2,
       text.col = c("#51af84","#b7371d"),
       legend   = c("HC", "COVID-19"),
       lty      = 0,
       cex      = 0.75)
```


The MDS plot presents variation amongst samples based on the normalised gene expression. The distances between each pair of samples on the MDS plot suggest their leading logFC. The healthy control (HC) and the patients (COVID-19) samples form separate clusters along the plot dimension 1. There are however still several COVID-19 samples mixed in the HC cluster.



```{r, warning=FALSE, message=FALSE, echo=FALSE}
# Regression model with group as target and samples as predictors
model_design <- model.matrix(~group, data  = samples_by_group)
dispersion   <- edgeR::estimateDisp(y      = gene_counts_DGE,
                                    design = model_design)
edgeR::plotBCV(
    y           = dispersion,
    col.tagwise = "black",
    col.common  = "red",
    main        = paste("Common and Tag-wise Dispersion vs",
                        "Expression Levels")
)
```



The BCV plot presents the estimated relative variability of expression between biological replicates; it illustrates the association between the biological CV and the average true gene abundance. The common dispersion line suggests all gene expression values vary by close to a BCV value of 0.5 amongst replicates.[@bcv]
The tagwise dispersion shows BCV values calculated individually for each gene. We observe that genes with higher true abundance (under the assumption that RNA-seq counts follow a negative binomial distribution) have lower BCV's than genes with lower abundance. This imples that the higher the true expression level of a gene has, the lower its variation in our samples.


```{r, warning=FALSE, message=FALSE, echo=FALSE}
edgeR::plotMeanVar(
    object                       = dispersion,
    show.raw.vars                = TRUE,
    show.tagwise.vars            = TRUE,
    NBline                       = TRUE,
    show.ave.raw.vars            = TRUE,
    show.binned.common.disp.vars = TRUE,
    main                         = paste(
        "Mean-variance Relationship for 14425 Genes of",
        "COVID-19 Group and Healthy Control Group"),
    cex.main                     = 0.8
)
```


The mean-variance plot presents the modelling of the mean-variance relationship for the normalised expression values, which are split into bins by overall expression level in the model. The grey points represent the pooled genewise variances. The blue points on the mean-variance plot represent the estimated genewise variances. The red crosses represent the pooled genewise variances, while the dark red crosses represent the the average of the pooled variances for each bin of genes plotted against the average expression level of the genes in the bin. The blue line shows the mean-variance relationship for a negative binomial model with common dispersion.[@edgeR_1][@edgeR_2] We obeserve that all types of variances fit well along the negative binomial line.


Note that this result can only be reproduced using *Bioconductor 3.11*; using other versions of Bioconductor will lead to a different mapping of gene names.
